src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -1482,74 +1482,74 @@ private class RsFnInferenceContext(
         }
     }
 
-    private fun RsPat.extractBindings(type: Ty) {
+    private fun RsPat.extractBindings(type: Ty, ignoreRef: Boolean = false) {
         when (this) {
             is RsPatWild -> {}
             is RsPatConst -> expr.inferTypeCoercableTo(type)
             is RsPatRef -> pat.extractBindings((type as? TyReference)?.referenced ?: TyUnknown)
             is RsPatRange -> patConstList.forEach { it.expr.inferTypeCoercableTo(type) }
             is RsPatIdent -> {
                 val patBinding = patBinding
-                val bindingType = if (patBinding.isRef) TyReference(type, patBinding.mutability) else type
+                val bindingType = if (patBinding.isRef && !ignoreRef) TyReference(type, patBinding.mutability) else type
                 ctx.writeBindingTy(patBinding, bindingType)
                 pat?.extractBindings(type)
             }
             is RsPatTup -> {
-                val types = (type as? TyTuple)?.types.orEmpty()
+                val (derefTy, mut) = type.stripReferences()
+                val types = (derefTy as? TyTuple)?.types.orEmpty()
                 for ((idx, p) in patList.withIndex()) {
-                    p.extractBindings(types.getOrElse(idx, { TyUnknown }))
+                    p.extractBindings(types.getOrElse(idx) { TyUnknown }.toRef(mut))
```

Comment:
Looks like that we need pass `mut != null` as the second argument
It should fix the following case:
```rust
fn main() {
    let p = (123, 123);
    let (x, ref y) = &p;
}
```
I this case `x` and `y` have `&i32` type but we infer that type of `y` is `&&i32`
<img width="333" alt="screen shot 2018-07-09 at 15 01 38" src="https://user-images.githubusercontent.com/2539310/42449362-0cb9c5d2-8389-11e8-851d-5e6d96810114.png">